File: src/main/java/com/github/shatteredsuite/scrolls/data/scroll/ScrollCrafting.java
package com.github.shatteredsuite.scrolls.data.scroll;

import static com.github.shatteredsuite.core.config.ConfigUtil.getIfValid;
import static com.github.shatteredsuite.core.config.ConfigUtil.getMaterialOrDef;
import static com.github.shatteredsuite.core.config.ConfigUtil.reflectiveSerialize;

import com.github.shatteredsuite.core.config.ConfigRecipe;
import com.github.shatteredsuite.core.include.xseries.XMaterial;
import java.util.Map;
import org.bukkit.Material;
import org.bukkit.configuration.serialization.ConfigurationSerializable;
import org.bukkit.configuration.serialization.SerializableAs;

@SerializableAs("ScrollCrafting")
public class ScrollCrafting implements ConfigurationSerializable {
    public final ConfigRecipe recipe;
    public final Material repairMaterial;
    public final int craftAmount;
    public final boolean craftable;
    public final boolean repairable;

    /**
     * Default to non-craftable.
     */
    public ScrollCrafting() {
        this.repairMaterial = null;
        this.recipe = null;
        this.craftAmount = 0;
        this.craftable = false;
        this.repairable = false;
    }

    /**
     * Constructor.
     *
     * @param recipe The recipe that should be used to craft this scroll.
     * @param repairMaterial The material that should be used to repair this scroll.
     * @param craftAmount The amount of scrolls to craft.
     */
    public ScrollCrafting(ConfigRecipe recipe, Material repairMaterial, int craftAmount) {
        this.recipe = recipe;
        this.repairMaterial = repairMaterial;
        this.craftAmount = craftAmount;
        this.craftable = recipe != null && recipe.valid && craftAmount > 0;
        this.repairable = repairMaterial != null && repairMaterial != XMaterial.AIR.parseMaterial();
    }

    /**
     * ConfigurationSerializable deserialization method.
     *
     * @param map The map to deserialize from.
     * @return the ScrollCrafting created from the map.
     */
    public static ScrollCrafting deserialize(Map<String, Object> map) {
        ConfigRecipe recipe = getIfValid(map, "recipe", ConfigRecipe.class, null);
        Material mat = getMaterialOrDef(map, "repair-material", XMaterial.ENDER_PEARL.parseMaterial());
        int craftAmount = getIfValid(map, "craft-amount", Integer.class, 1);
        return new ScrollCrafting(recipe, mat, craftAmount);
    }

    /**
     * ConfigurationSerializable serialization method.
     *
     * @return A serialized map of this.
     */
    @Override
    public Map<String, Object> serialize() {
        return reflectiveSerialize(this, ScrollCrafting.class);
    }
}

File: src/main/kotlin/com/github/shatteredsuite/scrolls/data/scroll/ScrollCrafting.kt
package com.github.shatteredsuite.scrolls.data.scroll

import com.github.shatteredsuite.core.config.ConfigRecipe
import com.github.shatteredsuite.core.include.xseries.XMaterial
import com.github.shatteredsuite.scrolls.ShatteredScrolls
import com.github.shatteredsuite.scrolls.items.ScrollInstance
import org.bukkit.Material
import org.bukkit.NamespacedKey

class ScrollCrafting {
    @JvmField
    val recipe: ConfigRecipe?
    @Transient
    var key: NamespacedKey? = null
        private set

    @JvmField
    val repairMaterial: Material?
    val repairAmount: Int
    val craftAmount: Int

    val craftable: Boolean
        get() = recipe != null && craftAmount > 0

    val repairable: Boolean
        get() = repairMaterial != null && repairAmount > 0 && repairMaterial != XMaterial.AIR.parseMaterial()

    /**
     * Default to non-craftable.
     */
    constructor() {
        repairMaterial = null
        recipe = null
        craftAmount = 0
        repairAmount = 0
    }

    /**
     * Constructor.
     *
     * @param recipe The recipe that should be used to craft this scroll.
     * @param repairMaterial The material that should be used to repair this scroll.
     * @param repairAmount The amount of charges to restore to the scroll.
     * @param craftAmount The amount of scrolls to craft.
     */
    constructor(recipe: ConfigRecipe?, repairMaterial: Material?, repairAmount: Int, craftAmount: Int) {
        this.recipe = recipe
        this.repairMaterial = repairMaterial
        this.repairAmount = repairAmount
        this.craftAmount = craftAmount
    }

    fun setKey(plugin: ShatteredScrolls, type: ScrollType) {
        this.key = NamespacedKey(plugin, type.id)
    }
}